src/windows.jl

```diff
@@ -25,150 +25,479 @@ export  rect,
         kaiser,
         dpss,
         dpsseig
+
+# this gets shared between the docstrings of all the window functions because
+# they share these args.
+zerophase_docs = """
+If `zerophase` is `false` (the default) the window is centered around index
+`(n+1)/2`, which is commonly used for FIR filter design. These are often used
+for FIR filter design, and usually odd-length. Note that for even-length windows
+this will cause the window center to be located between samples.
+
+If `zerophase` is `true` the window is centered around index 1 (with the
+negative half wrapped to the end of the vector). Additionally this creates a
+"periodic" window, which means that if there is no padding then the left and
+right endpoints of the window wrap around to the same sample, so the window
+length is the same as an `n+1`-length non-`zerophase` window. Alternatively you
+can think of the continuous `zerophase` window being of width `n` and the
+non-`zerophase` window being of length `n-1`. `zerophase` windows are often used
+in FFT processing, and are usually even-length.
+"""
+
+
+function twoD_docs(arg=nothing)
+    argstr = if arg !== nothing
+        "`$arg`, "
+    else
+        ""
+    end
+    """
+    Providing a 2D `Tuple` rather than a single `n` constructs a 2D window.
+    $argstr`padding` and `zerophase` can then be given either as a single value
+    for both horizontal and vertical or a 2-tuple to specify them separately.
+    """
+end
+
+"""
+    padplot(plotstr)
+
+Takes a multiline string and pre-pads it so that it shows up as
+preformatted code when included in a docstring.
+"""
+function padplot(plotstr)
+    minpad = typemax(Int)
+    lines = split(plotstr, "\n")
+    # count the minimum number of spaces preceeding any line
+    for line in lines
+        line == "" && continue
+        minpad = min(minpad, length(match(r"^ *", line).match))
+    end
+
+    pad = " " ^ (4-minpad)
+    join((string(pad, line) for line in lines), "\n")
+end
+
+include("winplots.jl")
+
+"""
+    makewindow(winfunc::Function, n, padding, zerophase)
+
+Generate a discrete window vector of the given length with `padding` zeros.
+`winfunc` should be a function giving the window value in the range of
+[-0.5, 0.5]. The function is assumed to be 0 outside of this range.
+
+$zerophase_docs
+
+Example:
+
+```
+function hanning(n::Integer; padding=0, zerophase=false)
+    makewindow(n, padding, zerophase) do x
+        0.5*(1+cos(2pi*x))
+    end
+end
+```
+"""
+function makewindow(winfunc::Function, n, padding, zerophase)
+    win = zeros(n+padding)
+    if n == 1
```

Comment:
It would be nice to validate inputs to this function, since most (all?) window functions now use it to generate windows. e.g. ensure that `padding >= 0` (negative values currently throw a cryptic error), or even that `n >=0`, and provide helpful error messages.

Reply:
Good catch - added some checks.